File: java/l2server/util/xml/XmlDocument.java
/*
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program. If not, see <http://www.gnu.org/licenses/>.
 */

package l2server.util.xml;

import l2server.log.Log;
import org.w3c.dom.Document;
import org.w3c.dom.Node;

import javax.xml.parsers.DocumentBuilderFactory;
import java.io.File;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Pere
 */
public class XmlDocument
{
	private static DocumentBuilderFactory _factory = DocumentBuilderFactory.newInstance();

	static
	{
		_factory.setValidating(false);
		_factory.setIgnoringComments(true);
	}

	private List<XmlNode> _children = new ArrayList<>();

	public XmlDocument(File file)
	{
		if (file == null || !file.exists())
		{
			Log.warning("The following XML could not be loaded:");
			Log.warning("- " + file.getAbsolutePath());
			return;
		}

		Document doc = null;
		try
		{
			doc = _factory.newDocumentBuilder().parse(file);
		}
		catch (Exception e)
		{
			e.printStackTrace();
		}

		for (Node baseNode = doc.getFirstChild(); baseNode != null; baseNode = baseNode.getNextSibling())
		{
			if (baseNode.getNodeType() == Node.ELEMENT_NODE)
			{
				_children.add(new XmlNode(baseNode));
			}
		}
	}

	public XmlNode getFirstChild()
	{
		if (_children.isEmpty())
		{
			return null;
		}

		return _children.get(0);
	}

	public List<XmlNode> getChildren()
	{
		return _children;
	}
}

File: src/main/java/l2server/util/xml/XmlDocument.kt
package l2server.util.xml

import org.w3c.dom.Node
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.InputStream
import javax.xml.parsers.DocumentBuilderFactory

/**
 * @author Pere
 */
class XmlDocument {

	val root: XmlNode

	constructor(file: File) {
		if (!file.exists()) {
			throw FileNotFoundException("The following XML could not be loaded: " + file.absolutePath)
		}

		val stream = FileInputStream(file)
		root = load(stream)
	}

	constructor(stream: InputStream) {
		root = load(stream)
	}

	fun getChildren(): List<XmlNode> = root.getChildren()

	private fun load(stream: InputStream): XmlNode {
		val doc = DOCUMENT_BUILDER_FACTORY.newDocumentBuilder().parse(stream)

		var baseNode: Node? = doc.firstChild
		while (baseNode != null) {
			if (baseNode.nodeType == Node.ELEMENT_NODE) {
				return XmlNode(baseNode)
			}
			baseNode = baseNode.nextSibling
		}

		throw RuntimeException("Tried to load an empty XML document!")
	}

	companion object {
		private val DOCUMENT_BUILDER_FACTORY = DocumentBuilderFactory.newInstance()

		init {
			DOCUMENT_BUILDER_FACTORY.isValidating = false
			DOCUMENT_BUILDER_FACTORY.isIgnoringComments = true
		}
	}
}
